src/mlpack/methods/ann/layer/neural_turing_machine_impl.hpp

```diff
@@ -0,0 +1,367 @@
+/**
+ * @file neural_turing_machine_impl.hpp
+ * @author Sumedh Ghaisas
+ *
+ * Implementation of NeuralTuringMachine layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_NTM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_NTM_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "neural_turing_machine_impl.hpp"
+
+#include "../visitor/forward_visitor.hpp"
+#include "../visitor/backward_visitor.hpp"
+#include "../visitor/gradient_visitor.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template <typename InputDataType, typename OutputDataType>
+NeuralTuringMachine<InputDataType, OutputDataType>::NeuralTuringMachine(
+    const size_t inSize,
+    const size_t outSize,
+    const size_t numMem,
+    const size_t memSize,
+    const size_t shiftSize,
+    LayerTypes controller) :
+    inSize(inSize),
+    outSize(outSize),
+    numMem(numMem),
+    memSize(memSize),
+    shiftSize(shiftSize),
+    controller(controller),
```

Comment:
You can use `controller(std::move(controller))` here.  See https://github.com/mlpack/mlpack/issues/1021 for a little more detail on why.

Reply:
You are right, but I am not sure how much performance difference it would make since LayerTypes is only holding the pointers.